### DISCUSSION TITLE
Hub zu Hub

### DISCUSSION THREAD

**Discussion by msw-home** [pybricks/support#1732] [Q&A]
Moin, ich suche ein einfaches Blockprogramm, um von Hub zu Hub, zu kommunizieren. Möchte am 1. Hub die Lego Fernbedienung anschließen um dann am 2. Hub, 2 Motoren ansteuern.
Grüße Manfred

**Comment by laurensvalk** [comment]
Here’s an example for inspiration: https://pybricks.com/project/technic-42146-powered-up-remote/

**Reply by msw-home** [reply]
Ich habe jetzt Masterhub und einen Slavehub definiert.
Die Übertragung funktioniert.
Im Masterhub habe ich die Anwahl über die Fernbedienung programmiert.
Im Slavehub ist die Ansteuerung für den Motor an Anschluss „A“.
Also Datenrichtung von Master nach Slave. Fernbedienung zum Motor, läuft ohne Probleme.

Jetzt meine Frage:
Kann man jetzt einen Farbsensor an den Anschluss „B“ des Slavehub anschließen
und die empfangenen Daten auf den Masterhub übertragen?
Also die umgekehrte Datenrichtung von Slave zu Master.

Gruß Manfred

![Master - pybricks_program_2024-07-31T16_12_49](https://github.com/user-attachments/assets/95afb4af-21ae-48de-be47-7de59db6c83c)
![Slave - pybricks_program_2024-07-31T16_11_41](https://github.com/user-attachments/assets/952430eb-eb5a-4c2f-8f00-fca1b6cf9883)

**Reply by laurensvalk** [reply]
Yes, you can make the communication bidirectional.

This works quite well on the SPIKE/Inventor hubs you are using. On Technic Hub, it is best to avoid bidirectional communication if possible due to limitations of its Bluetooth chip.

**Reply by msw-home** [reply]
Danke euch für die Hilfestellung, die Hub zu Hub Verbindung in beide Richtungen funktioniert nun.
Nächstes Problem, siehe Bild unten.
Fehlermeldung: The resourece cannot be used in two tasks at once
Gibt es hierfür eine Möglichkeit, die beiden Programmteile zusammen zubringen?

![image](https://github.com/user-attachments/assets/fd04b5fa-4937-4b19-b042-10e1a428f873)

**Reply by laurensvalk** [reply]
Yes. Changing the color of the remote light and broadcasting are both Bluetooth commands.

Instead of setting the light directly, you could store the color you want in a variable. Then add the block to set the remote color to the same loop where you broadcast your data.

Then the Bluetooth commands always happen after each other instead of at the same time.

**Reply by msw-home** [reply]
Sorry, das habe ich nicht verstanden, kannst Du mir das mal mit einer Farbe programmieren.

**Reply by laurensvalk** [reply]
Here's how you can do it. I have simplified the loops a little bit, to illustrate the main point:

![image](https://github.com/user-attachments/assets/f1a38226-54ad-4ec2-89b5-b10d0722d882)

**Reply by msw-home** [reply]
Hallo Laurens und Bert, ich habe die Hub zu Hub soweit hinbekommen, mit vielen Test, aber richtig immer noch nicht verstanden.
Dabei die beiden Programmteile, die als Test benutzt habe.
Aus meiner Sicht, keine Probleme.
Ich kann 2 Motoren und ein Farbsensor bedienen, bzw. die Werte abzufragen.
![gelber Hub Test Hub to Hub_pybricks_program_2024-08-03T15_33_21](https://github.com/user-attachments/assets/22557a41-0ef1-42c0-a8c1-efff113f1a81)
![grüner Hub Test Hub to Hub_pybricks_program_2024-08-03T15_33_34](https://github.com/user-attachments/assets/f49beec5-37b3-4bf1-82fc-b270218d5716)
Aber jetzt stoße ich mit meine Fähigkeiten an die Grenze!
Ich möchte das auf mein Hauptprojekt, übertragen, keine Change!
Das Hauptprojekt ein 6 axis-roboterarm, lauffähig mit nur einem Hub! Dadurch sind die 6 Schnittstellen, für die 6 Motoren belegt.
Ich möchte jetzt eine Erweiterung um weitere 6 Schnittstellen erreichen, des der zweite Hub.
Hierdrüber möchte ich 2 Motoren und einen Farbsensor integrieren.
![6-axis Roboter Arm_pybricks_program_2024-08-03T15_33_47](https://github.com/user-attachments/assets/ec69821b-9ce1-4d39-92e7-d7f24ed7da96)
Ist das überhaupt möglich, alles über eine Fernbedienung?
Wenn ja, wäre einer bereit mich zu unterstützen, wäre auch bereit bei Erfolg, eine entsprechende Spende zu leisten.

**Reply by BertLindeman** [reply]
Manfred, 

This is a bit more than I can grasp in a minute 😄 
My memory is not as sharp as it was, but I saw (and have impemented) a program that switches mode using the center button of the remote control. You might have a mode "green" to control  the 6 motors and a "yellow" mode to control the next 2 motors and a mode blue to use the color senor.

The example I know exists I cannot find now, but I'll keep on searching.

Good, that you are using pybricks. It's so good! 

Bert

**Reply by msw-home** [reply]
Hallo Bert,
ok kann ich nachvollziehen, ich kann warten, bin Rentner!
Die 6 Motoren vom Robo, werden über den Modus "grün" Achse 1-3 mit den Tasten "+"-center-"-" gesteuert.
Achse 4-6 im Modus "gelb" mit den Tasten "+"-center-"-" gesteuert.
Modus "rot" ist für Automatikabläufe geplant.
Modus "weiß" oder ein neuer Modus "blau" wäre für die 2 zusätzlichen Motore/Farbsensor angedacht.
Hierdrüber möchte ich ein Förderband steuern.
Viel Spaß und schönes Wochenende,
gut das es Pybricks gibt und macht richtig Spaß, --- wenn es funktioniert!
Aber jedes Programm ist nur so gut, wie der Anwender!

**Reply by BertLindeman** [reply]
Rentner? Wie ich.
Ich verstehe nicht, wie ich jemals Zeit hatte, um zu arbeiten.

**Reply by msw-home** [reply]
Moin zusammen, die meisten Fragen/Probleme habe ich gelöst.
Ich kann über beide Hub`s meine Motoren über eine Fernbedienung ansteuern.
Was noch nicht klappt, ist die LED Ansteuerung auf der FB.
Werde ich mir noch anschauen, also keine Aktivitäten, von euer Seite her mehr nötig.
Danke für die bisherige Unterstützung, es erden bestimmt noch andere Probleme auftauchen,
wenn ich meine Testumgebung in das Projekt einbinden werde.
LG Manfred

**Reply by msw-home** [reply]
Hallo zusammen,
bei der Ansteuerung der LED auf der Fernbedienung stoße ich an meine Grenzen.
Einer der beiden Hub`s streikt immer.
Hilfe wäre wünschenswert.
Ich möchte 4 Farben anzeigen, die mit der Centertaste durchgerollt werden, weiß - grün - gelb - rot
entspricht "Umschaltung FB Modus"
0 = weiß
1 = grün
2 = gelb
4 = rot

Anbei die beiden lauffähigen Projekt Programme.
![Master --- pybricks_program_2024-08-05T12_06_54](https://github.com/user-attachments/assets/2bec7d9f-e5c4-4cac-960d-23be502919ee)
![Slave --- pybricks_program_2024-08-05T12_07_13](https://github.com/user-attachments/assets/121bd237-c62e-401c-b631-86a1fb58185d)

**Comment by BertLindeman** [comment]
Manfred,

In fact quite the same but using another channel to send / receive. 
On the definition of the hubs you specify both the broadcast and observe channel. 

Bert
